File: magellan-library/src/test/java/com/wealthfront/magellan/ScreenTest.java
package com.wealthfront.magellan;

import android.app.Activity;
import android.app.Dialog;
import android.content.Context;
import android.os.Bundle;
import android.os.Parcelable;
import android.util.SparseArray;
import android.view.View;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.ArgumentCaptor;
import org.mockito.Captor;
import org.mockito.Mock;
import org.mockito.invocation.InvocationOnMock;
import org.mockito.stubbing.Answer;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.annotation.Config;

import static com.google.common.truth.Truth.assertThat;
import static org.mockito.Matchers.isA;
import static org.mockito.Mockito.doAnswer;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.MockitoAnnotations.initMocks;

@Config(manifest = Config.NONE)
@RunWith(RobolectricTestRunner.class)
public class ScreenTest {

  @Mock BaseScreenView<DummyScreen> view;
  @Mock Dialog dialog;
  @Captor ArgumentCaptor<SparseArray<Parcelable>> sparseArrayCaptor;
  private DummyScreen screen;

  @Before
  public void setUp() {
    initMocks(this);
    screen = new DummyScreen(view);
  }

  @Test
  public void recreateView() {
    View v = screen.recreateView(null, null);

    assertThat(screen.getView()).isEqualTo(v);
    assertThat(view).isEqualTo(v);
    assertThat(view.getScreen()).isEqualTo(screen);
  }

  @Test
  public void destroyView() {
    screen.recreateView(null, null);
    screen.destroyView();

    verify(view).saveHierarchyState(isA(SparseArray.class));
  }

  @Test
  public void createDialog() {
    screen.showDialog(new DialogCreator() {
      @Override
      public Dialog createDialog(Activity activity) {
        return dialog;
      }
    });
    screen.createDialog();

    verify(dialog, times(2)).show();
  }

  @Test
  public void destroyDialog() {
    screen.showDialog(new DialogCreator() {
      @Override
      public Dialog createDialog(Activity activity) {
        return dialog;
      }
    });
    screen.destroyDialog();

    verify(dialog).setOnDismissListener(null);
    verify(dialog).dismiss();
  }

  @Test
  public void saveRestore() {
    final Bundle state42 = new Bundle();
    state42.putString("key", "value");
    doAnswer(new Answer() {
      @Override
      public Object answer(InvocationOnMock invocation) throws Throwable {
        sparseArrayCaptor.getValue().put(42, state42);
        return null;
      }
    }).when(view).saveHierarchyState(sparseArrayCaptor.capture());

    screen.recreateView(null, null);
    Bundle bundle = new Bundle();
    screen.save(bundle);
    screen.restore(bundle);
    screen.recreateView(null, null);

    verify(view).saveHierarchyState(isA(SparseArray.class));
    verify(view).restoreHierarchyState(sparseArrayCaptor.capture());
    assertThat(((Bundle) sparseArrayCaptor.getValue().get(42)).getString("key")).isEqualTo("value");
  }

  private static class DummyScreen extends Screen<BaseScreenView<DummyScreen>> {

    private BaseScreenView<DummyScreen> view;

    DummyScreen(BaseScreenView<DummyScreen> view) {
      this.view = view;
    }

    @Override
    protected BaseScreenView<DummyScreen> createView(Context context) {
      return view;
    }

  }

}

File: magellan-library/src/test/java/com/wealthfront/magellan/core/ScreenTest.kt
package com.wealthfront.magellan.core

import android.os.Bundle
import android.os.Parcelable
import android.util.SparseArray
import android.view.LayoutInflater.from
import android.view.View
import com.google.common.truth.Truth.assertThat
import com.wealthfront.magellan.databinding.MagellanDummyLayoutBinding
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.ArgumentCaptor
import org.mockito.ArgumentMatchers.any
import org.mockito.Captor
import org.mockito.Mock
import org.mockito.Mockito.doAnswer
import org.mockito.Mockito.never
import org.mockito.Mockito.verify
import org.mockito.MockitoAnnotations.initMocks
import org.robolectric.RobolectricTestRunner
import org.robolectric.RuntimeEnvironment.application
import org.robolectric.annotation.Config

@Config(manifest = Config.NONE)
@RunWith(RobolectricTestRunner::class)
class ScreenTest {

  private lateinit var screen: DummyScreen
  @Captor lateinit var sparseArrayCaptor: ArgumentCaptor<SparseArray<Parcelable>>
  @Mock lateinit var view: View

  @Before
  fun setUp() {
    initMocks(this)
    screen = DummyScreen()
    screen.view = view
    screen.viewBinding = MagellanDummyLayoutBinding.inflate(from(application))
  }

  @Test
  fun onShow() {
    screen.show(application)

    verify(view, never()).restoreHierarchyState(any())
  }

  @Test
  fun onHide() {
    screen.hide(application)

    verify(view).saveHierarchyState(any())
  }

  @Test
  fun saveRestore() {
    val state = Bundle()
    state.putString("key", "value")
    doAnswer { sparseArrayCaptor.value.put(42, state) }.`when`(view).saveHierarchyState(sparseArrayCaptor.capture())

    screen.hide(application)
    verify(view).saveHierarchyState(any())

    screen.show(application)
    verify(view).restoreHierarchyState(sparseArrayCaptor.capture())

    val bundle = sparseArrayCaptor.value.get(42) as Bundle
    assertThat(bundle.getString("key")).isEqualTo("value")
  }
}

class DummyScreen : Screen<MagellanDummyLayoutBinding>(MagellanDummyLayoutBinding::inflate)
